Support for an empty environment

pdcp1 opened the pull request:
This is part of the issue https://github.com/responsible-ai-collaborative/aiid/issues/2224

## New environment and site
- New Atlas environment: [aiid-dev-empty](https://cloud.mongodb.com/v2/64f74f2c486ff808bae5cc32)
- New Netlify site: [aiid-dev-empty](https://app.netlify.com/sites/aiid-dev-empty/overview)

## This PR includes
- Create typeDefs file with all Gatsby schema customization
- Add an optional`IS_EMPTY_ENVIRONENT` env variable. It should be `true` only for an empty environment.
- Add `conditionalIt` Cypress function to check conditions to run tests.
- Add a new Github Action `realm-empty-env.yml` that runs automatically when we push to staging branch and deploy the staging code to the Empty Environment Atlas app described above (`aiid-dev-empty`)
- Add 3 new Github Action secrets:
  - `GATSBY_REALM_EMPTY_ENV_APP_ID` = The Realm App ID of the empty environment
  - `REALM_EMPTY_ENV_API_PRIVATE_KEY` = The private key of the empty environment
  - `REALM_EMPTY_ENV_API_PUBLIC_KEY` = The public key of the empty environment

## Preview link for testing
https://feature-empty-env--aiid-dev-empty.netlify.app/

cesarvarela reviewed (commented):
Do you have a deploy preview of this?

We could merge this as it is, but I think the empty test suite should ensure that the user's first experience with the app should test that new incidents, reports, etc, can be created successfully and nothing crashes.

pdcp1 commented:
> Do you have a deploy preview of this?

@cesarvarela Yes, I just added it to the PR description. I also added the new Netlify site and Atlas env links.

pdcp1 commented:
@cesarvarela I'm thinking what is the best way to create a default user (it should have `admin` role to test the whole empty system).
We can create it manually through the Atlas `App Users` UI and then go to the DB and change the `subscriber` role to `admin`, and then update the ENV variables `E2E_ADMIN_USERNAME` and `E2E_ADMIN_PASSWORD`.

But I'm thinking of a more automated way to do this. Any ideas?

pdcp1 commented on site/gatsby-site/cypress/e2e/integration/pages.cy.js:
```diff
@@ -151,7 +154,7 @@ describe('Pages', () => {
   });
 
-  it('Should generate a valid RSS feed', () => {
+  conditionalIt(!Cypress.env('isEmptyEnvironment'), 'Should generate a valid RSS feed', () => {
```
I moved RSS tests to a separate `rss.cy.js` file. Since we need separate scenarios to test empty XML and XML with actual data.

pdcp1 commented on site/gatsby-site/cypress/e2e/integration/discover.cy.js:
```diff
@@ -12,23 +13,27 @@ describe('The Discover app', () => {
 
-  it('Should default to incident reports and show at least 30', () => {
-    cy.visit(url);
+  conditionalIt(
```
Added

pdcp1 commented on site/gatsby-site/cypress/e2e/integration/wordcounts.cy.js:
```diff
@@ -1,4 +1,8 @@
 describe('The Word Counts Page', () => {
+  before('before', function () {
```
Done

pdcp1 commented on site/gatsby-site/cypress/e2e/integration/landingPage.cy.js:
```diff
@@ -57,7 +57,7 @@ describe('The Landing page', () => {
   });
 
-  it('Should display common entities card', () => {
+  conditionalIt(!Cypress.env('isEmptyEnvironment'), 'Should display common entities card', () => {
```
Added

pdcp1 commented on site/gatsby-site/cypress/e2e/integration/entity.cy.js:
```diff
@@ -11,6 +11,10 @@ const entity = {
 const USER_ID = '63320ce63ec803072c9f529c';
 
 describe('Entities page', () => {
+  before('before', function () {
```
Done

pdcp1 commented on site/gatsby-site/cypress/e2e/integration/integrity.cy.js:
```diff
@@ -14,6 +14,10 @@ const isLinked = (reportNumber, incidents) => {
 };
 
 describe('Integrity', () => {
+  before('before', function () {
```
They don't pass because they use `cy.query` that requires an admin user to authenticate the Apollo query.

pdcp1 replied:
As soon the default user is created, we can remove the skip here

pdcp1 commented on site/gatsby-site/cypress/e2e/integration/report.cy.js:
```diff
@@ -1,4 +1,8 @@
 describe('Report pages', () => {
+  before('before', function () {
+    Cypress.env('isEmptyEnvironment') == 'true' && this.skip();
```
This `describe` tests an specific report page (`/reports/2302`) so it can't be tested since those report pages are not generated yet.

pdcp1 commented on site/gatsby-site/cypress/e2e/integration/subscriptions.cy.js:
```diff
@@ -11,6 +11,13 @@ const entitySubscriptions = subscriptionsData.data.subscriptions
   .sort((a, b) => a.entityId.name - b.entityId.name);
 
 describe('Subscriptions', () => {
+  before('before', function () {
```
The subscriptions (account) page needs at least a default user. That's why I skip all of them until we have the default user set.

pdcp1 commented on site/gatsby-site/cypress/e2e/integration/apps/classifications.cy.js:
```diff
@@ -5,6 +5,10 @@ import classifications from '../../../fixtures/call/classifications.json';
   const url = '/apps/classifications';
 
+  before('before', function () {
+    Cypress.env('isEmptyEnvironment') == 'true' && this.skip();
```
I changed this to use a real Boolean variable here: https://github.com/responsible-ai-collaborative/aiid/blob/2e8932b5911ed0562dcfb800759ba958500d7044/site/gatsby-site/cypress/plugins/index.js#L24

pdcp1 commented on site/gatsby-site/cypress/e2e/integration/apps/classifications.cy.js:
```diff
@@ -5,6 +5,10 @@ import classifications from '../../../fixtures/call/classifications.json';
 describe('Classifications App', () => {
   const url = '/apps/classifications';
 
+  before('before', function () {
```
Done

cesarvarela commented:
It should be possible using this:

https://www.mongodb.com/docs/atlas/app-services/admin/api/v3/#tag/users/operation/adminCreateUser

Have an Atlas function that checks if this user is created since we already have auth credentials for the API there.

I guess we should add that in another PR 🤔

pdcp1 commented:
@cesarvarela I just created a new [issue](https://github.com/responsible-ai-collaborative/aiid/issues/2329) to resolve the Default Admin user creation. If you are OK with this PR, we can proceed to merge this first and then work on the new one

pdcp1 commented:
@kepae In order to deploy the site on the Atlas app, we need to create a public and private key. Please follow this [documentation](https://www.mongodb.com/docs/atlas/configure-api-access/#std-label-create-org-api-key) and let me know the values.
Thanks!

kepae commented:
@pdcp1 please let me know if this is ready for review and when you would be ready to merge it.

pdcp1 commented:
@kepae @cesarvarela This is ready for review.

Kevin, in addition to the code, we need to add 3 Github Action secrets on this repo:

- `GATSBY_REALM_EMPTY_ENV_APP_ID` = `aiidstitch2-paofd`
- `REALM_EMPTY_ENV_API_PUBLIC_KEY`= `mpqivflg`
- `REALM_EMPTY_ENV_API_PRIVATE_KEY` = The private key of the empty environment, the one that you send me on Slack

To add the Github Action secrets, follow these steps:
- Go to Settings
- Secrets and variables
- Actions
- Add the secret variables

kepae commented:
Thanks @pdcp1. Since this is a substantial code change, I'll prefer to update the secrets and merge this after the current production merge (https://github.com/responsible-ai-collaborative/aiid/pull/2325).

kepae commented on README.md:
```diff
@@ -440,8 +444,13 @@ These environments must contain the following variables:
 GATSBY_REALM_APP_ID=
 REALM_API_PRIVATE_KEY=
 REALM_API_PUBLIC_KEY=
+GATSBY_REALM_EMPTY_ENV_APP_ID=[The Realm App ID of the empty environment]
```
Are these env variables needed for the production and staging workflows? If not, I would suggest separating this into a smaller subsection, describing the empty environment with the new `realm-empty-env.yml` workflow.

pdcp1 replied:
Changed to use the `empty` environment variables instead of using these `EMPTY_DEV` ones

kepae replied:
is `GATSBY_REALM_EMPTY_ENV_APP_ID` still needed?

pdcp1 replied:
No, I just remove them from the README file 👍

kepae commented on .github/workflows/realm-empty-env.yml:
```diff
@@ -0,0 +1,26 @@
+
+on: [workflow_dispatch]
+
+#on:
```
commented code able to be removed?
I understand we want to keep this on manual workflow_dispatch and not run automatically on branch pushes.

kepae commented on site/gatsby-site/cypress/plugins/index.js:
```diff
@@ -21,6 +21,7 @@ module.exports = (on, config) => {
   config.env.e2eUsername = process.env.E2E_ADMIN_USERNAME;
   config.env.e2ePassword = process.env.E2E_ADMIN_PASSWORD;
   config.env.realmAppId = process.env.GATSBY_REALM_APP_ID;
+  config.env.isEmptyEnvironment = process.env.IS_EMPTY_ENVIRONMENT == 'true';
```
just double-checking -- this should be a string compare? the ENV var isn't read as a boolean, or?

pdcp1 replied:
Yes, this should be a string compare because the process.env variable is a string. I prefer to do this comparison here and use the `isEmptyEnvironment` setting as a boolean in the rest of the code

kepae reviewed (approved):
Few quick questions, then we should merge this so it doesnt go stale. :-)
